server/src/main/java/org/eclipse/lsp/cobol/core/model/tree/ProgramNode.java

```diff
@@ -44,6 +44,8 @@
 public class ProgramNode extends Node {
   private final Multimap<String, VariableNode> variables = ArrayListMultimap.create();
   private final List<CodeBlockDefinitionNode> codeBlocks = new ArrayList<>();
+  private final Map<String, List<ParagraphDefinitionNameNode>> paragraphNodes = new HashMap<>();
```

Comment:
There is no need to have a list of `ParagraphDefinitionNameNode`. You can use only one definition instance and collect all links there.
Right now in `registerParagraphUsage` you iterate through definitions and add usage to each. So that means you will have several copies of the same definitions. Use only one definition for the particular name.

Reply:
I updated the logic to retrieve definitions.